### PR TITLE
Custom paymaster deep copy edit and instruction flow check

### PR DESCRIPTION
**Pull request by kailash-manasarovar**
* All steps checked.
* Code corrected where necessary.
* Missing steps included.
* User's addition added.
* Flow tested and working with latest compiler.

### CODE REVIEW

File: docs/dev/tutorials/custom-paymaster-tutorial.md [hunk 1]
```diff
@@ -1,72 +1,101 @@
...
+    },
+  },
+  solidity: {
+    version: "0.8.8",
+  },
+};
+```
 
+::: tip
+- Use [npx hardhat](https://hardhat.org/hardhat-runner/docs/getting-started#quick-start) to scaffold a default Hardhat project.
```
Comment (uF4No) [inline docs/dev/tutorials/custom-paymaster-tutorial.md:84]: I'd not advise people to use `npx hardhat`. At least with zkSync CLI you get all our dependencies but `npx hardhat` give you nothing that works with zkSync out of the box

Comment (kailash-manasarovar) [reply docs/dev/tutorials/custom-paymaster-tutorial.md:84]: OK, can you make the suggestion and I'll accept it?

File: docs/dev/tutorials/custom-paymaster-tutorial.md [hunk 2]
```diff
@@ -192,24 +225,25 @@ try
...
-
-That is why it is important to verify that the user provided all the allowed prerequisites to the transaction _before_ performing any logic. This is the reason we _first_ check that the user provided enough allowance, and only then do we do `transferFrom`.
+::: tip Validate all requirements first
+The [validation steps](../developer-guides/aa.md#the-validation-step) ensure that the paymaster won't throttle if the first storage read which has a different value from the execution on the API is a storage slot that belongs to the user.
 
+This is why it is important to verify transaction prerequisites _before_ performing any logic and why we _first_ check that the user provided enough allowance before calling `transferFrom`.
 :::
 
-### Full code of the paymaster
+## Create paymaster contract
```
Comment (uF4No) [inline docs/dev/tutorials/custom-paymaster-tutorial.md:234]: I find it confusing that we've gone through all the logic of the paymaster and after all that we're going to create it here.

File: docs/dev/tutorials/custom-paymaster-tutorial.md [hunk 3]
```diff
@@ -1,72 +1,101 @@
...
 
 ```sh
-yarn add @matterlabs/zksync-contracts @openzeppelin/contracts @openzeppelin/contracts-upgradeable
+mkdir contracts deploy
 ```
 
-Then create the `hardhat.config.ts` config file, `contracts` and `deploy` folders, like in the [quickstart tutorial](../building-on-zksync/hello-world.md).
-
-::: tip
+5. Create the file `hardhat.config.ts` and add the following, replacing the `ethNetwork` with an RPC URL:
```
[the remark this answers is not in the document]
Comment (kailash-manasarovar) [reply docs/dev/tutorials/custom-paymaster-tutorial.md:56]: OK, cool, @uF4No can you make the suggestion and I'll accept please?

File: docs/dev/tutorials/custom-paymaster-tutorial.md [hunk 4]
```diff
@@ -351,12 +383,13 @@ contract MyERC20 is ERC20 {
         return _decimals;
     }
 }
-
 ```
 
-## Deploying the paymaster
+## Compile and deploy the contracts
+
+The script below deploys the ERC20 contract and the paymaster contract. It also creates an empty wallet and mints some `MyERC20` tokens to use for the paymaster to use at a later step.
```
Comment (uF4No) [inline docs/dev/tutorials/custom-paymaster-tutorial.md:390]: typo at the end of this line

File: docs/dev/tutorials/custom-paymaster-tutorial.md (hunk 2 above)
Comment (kailash-manasarovar) [reply docs/dev/tutorials/custom-paymaster-tutorial.md:234]: **Create** is an action. Going through the logic is passive learning. The instruction below is: 

Create the `contracts/MyPaymaster.sol` file and copy/paste the following:

Comment (kailash-manasarovar) [reply docs/dev/tutorials/custom-paymaster-tutorial.md:234]: I'll change it a bit.

File: docs/dev/tutorials/custom-paymaster-tutorial.md (hunk 4 above)
Comment (kailash-manasarovar) [reply docs/dev/tutorials/custom-paymaster-tutorial.md:390]: done

File: docs/dev/tutorials/custom-paymaster-tutorial.md (hunk 1 above)
Comment (kailash-manasarovar) [reply docs/dev/tutorials/custom-paymaster-tutorial.md:84]: DONE

**Review by uF4No — DISMISSED** [review]

File: docs/dev/tutorials/custom-paymaster-tutorial.md [hunk 5]
```diff
@@ -1,72 +1,100 @@
...
-- Create the ERC20 token contract and send some tokens to a brand new wallet.
-- Finally we will send a `mint` transaction from the newly created wallet via the paymaster. Even though the transaction would normally require some ETH to pay for the gas fee, our paymaster will execute the transaction in exchange for 1 unit of the ERC20 token.
+- Create a paymaster that assumes a single unit of an ERC20 token is enough to cover any transaction fee.
+- Create the ERC20 token contract and send some tokens to a new wallet.
+- Send a `mint` transaction from the newly created wallet via the paymaster. Although the transaction normally requires ETH to pay the gas fee, our paymaster executes the transaction in exchange for 1 unit of the ERC20 token.
 
+## Prerequisites
 
-## Prerequisite
+- A [Node](https://nodejs.org/en/download) installation running Node 18.
```
[the remark this answers is not in the document]
Comment (kailash-manasarovar) [reply docs/dev/tutorials/custom-paymaster-tutorial.md:11]: I think we should say 16 to be safe.

File: docs/dev/tutorials/custom-paymaster-tutorial.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kailash-manasarovar) [reply docs/dev/tutorials/custom-paymaster-tutorial.md]: this suggestion didn't work for some reason, so i'm manually adding it now ...

**Review by uF4No — APPROVED** [review]

**Pull request merged**